### ISSUE TITLE
find() is not adding _type to filter or key condition expressions

### ISSUE THREAD

**Issue by rjmackay** [sensedeep/dynamodb-onetable#297]
**Describe the bug**
When using find() on a model which shares a PK with other models, `_type` is not used to construct a `sk` filter and so find is returning models of other types. Previously, this worked because _type was always added to `FilterConditions`.

The issue seems to be related to the `where` clause: with the where clause it fails, without it a sk filter is added and works.

**To Reproduce**

Our schema looks like:
```
    Consumer: {
      pk: { type: String, value: 'consumer#${consumerId}' },
      sk: { type: String, value: '${_type}' },
      name: { type: String }
    },
    Order: {
      pk: { type: String, value: 'consumer#${consumerId}' },
      sk: { type: String, value: '${_type}#${orderId}' },
      deletedAt: { type: Date }
    }
```

The query I'm running looks like:
```
const orders = await OrderModel.find(
    { consumerId },
    { log: true, where: 'attribute_not_exists(${deletedAt})' }
  )
```

Logs:
```
    info OneTable result for "find" "Order" {
      cmd: {
        ExpressionAttributeNames: { '#_0': 'deletedAt', '#_1': 'pk', '#_2': 'consumerId' },
        ExpressionAttributeValues: {
          ':_0': { S: 'consumer#01FXGP7ZRJ97Y090JJX9K5VJ8J' },
          ':_1': { S: '01FXGP7ZRJ97Y090JJX9K5VJ8J' }
        },
        FilterExpression: '(attribute_not_exists(#_0)) and (#_2 = :_1)',
        KeyConditionExpression: '#_1 = :_0',
        TableName: 'toutatis',
        ConsistentRead: false,
        ScanIndexForward: true
      },
      items: [
        {
          consumerId: '01FXGP7ZRJ97Y090JJX9K5VJ8J',
          name: 'test',
          createdAt: 2022-03-06T22:40:17.682Z,
          updatedAt: 2022-03-06T22:40:17.682Z
        },
        {
          consumerId: '01FXGP7ZRJ97Y090JJX9K5VJ8J',
          orderId: '01FXGP7ZSMCWK0VG8EXZ9C714T',
          addressId: '01FXGP7ZRDZFGF9S87ZDVPTXCC',
          originalAddressId: '01FXGP7ZRDZFGF9S87ZDVPTXCC',
          status: 'created',
          updatedAt: 2022-03-06T22:40:17.716Z,
          createdAt: 2022-03-06T22:40:17.716Z
        }
      ],
      op: 'find',
      properties: {
        pk: 'consumer#01FXGP7ZRJ97Y090JJX9K5VJ8J',
        consumerId: '01FXGP7ZRJ97Y090JJX9K5VJ8J'
      },
      params: {
        parse: true,
        high: true,
        log: true,
        where: 'attribute_not_exists(${deletedAt})',
        checked: true
      }
    }
```

**Environment (please complete the following information):**
 - Node 14
 - OneTable Version: 2.3.1
 - TypeScript Version: 4.5.5

**Comment by mobsense** [comment]
Hi Robbie.  Can you check with the repo code.

It works for us here. The code below in addContext should be adding the "type" and then that should be used in runTemplates.

```
        if (!this.generic) {
            properties[this.typeField] = this.name
        }
```

**Comment by mobsense** [comment]
Robbie, any thoughts I this?  I'd like to resolve or close until you can provide a bit more info.

**Comment by rjmackay** [comment]
@mobsense I've added a test case here: https://github.com/veho-technologies/dynamodb-onetable/pull/new/debug-297

```typescript
/*
    debug.ts - Just for debug

    Edit your test case here and invoke via: "jest debug"

    Or run VS Code in the top level directory and just run.
 */
import {AWS, Client, Entity, Match, Model, Table, print, dump, delay} from './utils/init'
import { OneSchema } from '../src/index.js'

jest.setTimeout(7200 * 1000)

//  Change with your schema
const schema = {
    version: '0.0.1',
    indexes: {
        primary: { hash: 'pk', sort: 'sk' },
        gs1: { hash: 'gs1pk', sort: 'gs1sk', project: 'all' },
    },
    models: {
      Consumer: {
        pk: { type: String, value: 'consumer#${consumerId}' },
        sk: { type: String, value: '${_type}' },
        consumerId: { type: String },
        name: { type: String }
      },
      Order: {
        pk: { type: String, value: 'consumer#${consumerId}' },
        sk: { type: String, value: '${_type}#${orderId}' },
        consumerId: { type: String },
        orderId: { type: String },
        deletedAt: { type: Date }
      }
    }
}

//  Change your table params as required
const table = new Table({
    name: 'DebugTable',
    client: Client,
    schema,
    logger: true,
})

type ConsumerType = Entity<typeof schema.models.Consumer>
type OrderType = Entity<typeof schema.models.Order>

//  This will create a local table
test('Create Table', async() => {
    if (!(await table.exists())) {
        await table.createTable()
        expect(await table.exists()).toBe(true)
    }
})

test('Test', async() => {
  const ConsumerModel = table.getModel<ConsumerType>('Consumer')
  const OrderModel = table.getModel<OrderType>('Order')

  await ConsumerModel.create({ consumerId: "test", name: 'Test' })
  await OrderModel.create({ consumerId: "test", orderId: "order1" })
  await OrderModel.create({ consumerId: "test", orderId: "order2" })

  const orders = await OrderModel.find(
    { consumerId: "test" },
    { log: true, where: 'attribute_not_exists(${deletedAt})' }
  )

  expect(orders).toHaveLength(2)
})

test('Destroy Table', async() => {
    await table.deleteTable('DeleteTableForever')
    expect(await table.exists()).toBe(false)
})
```

**Comment by mobsense** [comment]
Found the cause.

In runTemplate, the code below strips "${" portions of the sk when doing a find. If you hard coded your _type value, this would preserve the type field in the SK.

The value.replace() needs to be fixed to preserve ${_type} where _type may be modified via the Table constructor.

```
        /*
            Consider unresolved template variables. If field is the sort key and doing find,
            then use sort key prefix and begins_with, (provide no where clause).
         */
        if (value.indexOf('${') >= 0 && index) {
            if (field.attribute[0] == index.sort) {
                if (op == 'find' && !params.where) {
                    //  Strip from first ${ onward and retain fixed prefix portion
                    let typeField = this.typeField
                    value = value.replace(/\${*/g, '')
                    if (value) {
                        return {begins: value}
                    }
                }
            }
```

**Comment by mobsense** [comment]
Found and fixed.

Thanks for the test case.

The core issue was using a where clause. Previously, we disabled the synthesis of a begins_with sort key on the type if a where clause was provided.

**Comment by mobsense** [comment]
Fixed in 2.3.5
